### PR TITLE
SLEEP-1337 Add support for mobile bulk uploads

### PR DESCRIPTION
**Pull request by bramj**
In Trypelim, we have performance issues on the form uploads.
To improve the performance of the uploads, we're grouping forms and org units into zip files (batches of 2000 forms) and upload these with only 1 API call per zip file.

This PR adds a job to process these bulk uploads:
1. A new API endpoint `/api/mobile/bulkupload/` that takes a `zip_file`. This endpoint simply uploads the file to S3, launches an async job, and returns HTTP 204.
2. The process job creates a record in the `APIImport` for future reference (and to be able to tag as `has_problem`) and processes all files in the zip in a db transaction.

Related JIRA tickets : https://bluesquare.atlassian.net/browse/SLEEP-1337

## Self proofreading checklist

- ~~[ ] Did I use eslint and black formatters~~
- [x] Is my code clear enough and well documented
- ~~[ ] Are my typescript files well typed~~
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

image

### CODE REVIEW

File: iaso/tasks/process_mobile_bulk_upload.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                fi.file = File(f)
+                fi.instance_id = instance.id
+                fi.name = instance_file.name
+                fi.save()
+                instance_files.append(fi)
+
+    return instance_files
+
+
+# Trypelim-specific
```
Comment (bramj) [inline iaso/tasks/process_mobile_bulk_upload.py:120]: Main issue I see for merging this is how this code is Trypelim specific (although it won't affect any form that doesn't have a `serie_id` attribute)

**Comment by bramj** [comment]
@mathvdh assigning you for review, I think Martin is too busy at the moment

File: iaso/api/mobile/bulk_uploads.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+logger = logging.getLogger(__name__)
+
+
+class ZipFileSerializer(serializers.Serializer):
+    zip_file = serializers.FileField()
+
+    def validateZipFile(self):
+        if self.is_valid():
+            return self.validated_data["zip_file"]
+        raise ValueError("Zip file not valid")
```
Comment (madewulf) [inline iaso/api/mobile/bulk_uploads.py:30]: Does this do something for real, outside of checking that there is a file?

File: iaso/tasks/process_mobile_bulk_upload.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    the_task = task
+    the_task.report_progress_and_stop_if_killed(
+        progress_value=0,
+        progress_message=_("Starting"),
+        end_value=100,
+    )
+    user = User.objects.get(id=user_id)
+    project = Project.objects.get(id=project_id)
+
+    api_import = APIImport.objects.create(
```
Comment (madewulf) [inline iaso/tasks/process_mobile_bulk_upload.py]: I would have said that this should be the very first thing happening with the zip file, before any check on the user or other treatments. The goal is to store the data before any other bug can happen, and there, it seems to happen relatively late (after even the pickup by the worker if I understand correctly, which means that if the worker is off, the payload can be lost?)

File: iaso/tasks/process_mobile_bulk_upload.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        api_import.exception = format_exc()
+        api_import.save()
+        raise e
+
+    the_task.report_success_with_result(
+        message=result_message(user, project, start_date, start_time, stats),
+    )
+
+
+def process_instance_xml(directory, zip_ref, user):
```
Comment (madewulf) [inline iaso/tasks/process_mobile_bulk_upload.py:99]: Isn't this duplicated code from /sync/views.py 

In this case, we should make it a function as to ensure that when we update the code somewhere with new features, we don't forget to update it everywhere

**Review by madewulf — CHANGES_REQUESTED** [review]

File: iaso/api/mobile/bulk_uploads.py (hunk 2 above)
Comment (bramj) [reply iaso/api/mobile/bulk_uploads.py:30]: No, just a check of the file

File: iaso/tasks/process_mobile_bulk_upload.py (hunk 3 above)
Comment (bramj) [reply iaso/tasks/process_mobile_bulk_upload.py]: The file is uploaded to S3 before being passed to the worker, so we wouldn't really lose the data. That said, you're totally right we should create the import object first, which will make it a lot easier to track when something goes wrong. I'll fix it.

File: iaso/tasks/process_mobile_bulk_upload.py (hunk 4 above)
Comment (bramj) [reply iaso/tasks/process_mobile_bulk_upload.py:99]: :+1: agreed, I've extracted a function

**Review by madewulf — APPROVED** [review]

**Pull request merged**